File: Solutions/Week 4/lru-cache.py
class LRUCache:

    def __init__(self, capacity: int):
        self.h = {}
        self.bot = None
        self.top = None
        self.cap = capacity

    def get(self, key: int) -> int:
        if key not in self.h:
            return -1
        node = self.h[key]
        if node.top:
            node.top.bot = node.bot
        if node.bot:
            if node.top:
                node.bot.top = node.top            
        self.top = node
        if self.bot==node and node.top:
            self.bot = node.top
        self.top.top = None
        self.bot.bot = None
        print(self.h.keys(), self.bot.key, self.top.key)
        return self.top.val

    def put(self, key: int, value: int) -> None:
        if key in self.h:
            node = self.h[key]
            node.val = value
            node.key = key
        else:
            node = ListNode(value)  
            node.key = key

        if not self.top:
            self.top = node
            self.bot = node
        elif self.top:
            self.top.top = node
            node.bot = self.top
            self.top = node
        if key not in self.h and len(self.h)==self.cap:
            k = self.bot.key
            print(key, self.h.keys(), self.bot.key, self.top.key)
            self.bot = self.bot.top
            if not self.bot:
                self.bot = node
            if k not in self.h:
                return
            print(k,self.h[k].val)
            del self.h[k]
            print(key, self.h.keys(), self.bot.key, self.top.key)


        self.h[key] = node
        self.top = node
        self.top.top = None
        self.bot.bot = None

class ListNode:
    
    def __init__(self, val):
        self.val = val
        self.key = None
        self.top = None
        self.bot = None

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: Solutions/Week 4/lru-cache.java
class LRUCache {
    
    private HashMap<Integer, Node> hs;
    private Integer start;
    private Integer end;
    private int cap;

    public LRUCache(int capacity) {
        if (hs==null){
            hs = new HashMap<Integer, Node>();
        }
        cap = capacity;
    }
    
    public int get(int key) {
        if (!hs.containsKey(key)) {
            return -1;
        }
        Node n = hs.get(key);
        update(n);
        return n.val;
    }
    
    public void put(int key, int value) {
        Node ins = hs.getOrDefault(key, new Node(value));
        ins.key = key;
        ins.val = value;
        hs.put(key, ins);

        if (start==null) {
            start = key;
            end = key;
        }
        else {
            update(ins);
        }
    }
    
    private void update(Node n){     
        if (n.left!=null 
            && n.right!=null
            && hs.containsKey(start)){
            n.left.right = n.right;
            n.right.left = n.left;
        }
        updateStart(n);
        updateEnd();
    }
  
    private void updateStart(Node n){
        
        if (n.key.equals(end)){
            if (n.right!=null){
                end = n.right.key;
                n.right.left = null;
            }
        }
        if(!start.equals(n.key) && !n.key.equals(end)){
            Node prev = hs.get(start);
            start = n.key;
            prev.right = n;
            n.left = prev;
            n.right = null;
            
        }
    }
    
    private void updateEnd(){
        if (hs.size()>cap){
            Node prev = hs.get(end);
            Node n = prev.right;
            hs.remove(end);
            end = n.key;
            n.left = null;
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

class Node {
    int val = 0;
    Integer key;
    Node left;
    Node right;
    Node(int val){
        this.val = val;
    }
}